File: proj8/proj_8_testers/tests/test11.ml
# Pass-by-reference in a recursion call

begin
    fib(int x, int ref y) begin
        if x <= 1 then begin
            y := x;
        end else begin
            int a, b;
            fib(x - 1, a);
            fib(x - 2, b);
            y := a + b;
        end
    end

    int f := 8;
    int x;
    fib(f, x);
    write(x, "\n");
end

File: archive/proj8/proj_8_testers/c_files/test11.c
#include <stdio.h>

// Testing that you can pass-by-reference in a recursion call.
void fib(int x, int* y) {
  if (x <= 1) {
    *y = x;
  }
  else {
    int a, b;
    fib(x-1, &a);
    fib(x-2, &b);
    *y = a + b;
  }
}

int main() {
  int f = 8;
  int x;
  fib(f, &x);
  printf("%d\n", x);
  return 0;
}
